### PR TITLE
Add helper functions for metric conversion [awsecscontainermetricsreceiver]

### PR DESCRIPTION
**Pull request by hossain-rayhan**
**Description:** <Describe what has changed.>
This change adds helper functions for converting ECS resources metrics to OT metrics. 

**Link to tracking Issue:**
#457 

**Testing:**
Unit test added.

**Documentation:** 
README.md

### CODE REVIEW

File: receiver/awsecscontainermetricsreceiver/go.mod
Comment (asuresh4) [inline receiver/awsecscontainermetricsreceiver/go.mod:11]: Looks like these are from importing`github.com/golang/protobuf/ptypes/timestamp` and `google.golang.org/protobuf/types/known/timestamppb`. Do you need both? Could probably just use `timestamppb` in both places.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/utils.go
Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/utils.go:29]: That's a good idea.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go
Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:35]: AWS Observability Collector-> Amazon distribution of OpenTelemetry.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go
Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go:49]: Hi @asuresh4 , I am little bit confused here. Can you explain a bit more? These labels describes the properties to differentiate the metrics. Then when/what should I set as metric labels. Also, these labels are supposed to be converted into metric dimensions.

Comment (asuresh4) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go:49]: The [kubelet_stats receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/receiver/kubeletstatsreceiver/kubelet/accumulator.go) should be good example to understand the concept. It collects metrics from different types of resources (containers, pods, nodes). The properties of resources are added as labels on the resource. An exporter exporting the metric would treat labels from the resource and the metric as dimensions.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go
Comment (anuraaga) [inline receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go:21]: @asuresh4 @bogdandrutu Are metrics receivers still using opencensus proto?

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go
Comment (anuraaga) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:35]: FWIW, resource type doesn't exist in OTel protocol but is there right now for metrics since it still seems to use opencensus. So this value will generally be dropped

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go
Comment (anuraaga) [inline receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:34]: The service name corresponds with an application, not a backend, so for example `AuthService`, `SearchFrontend`, etc. We could fill this in with the ECS service name, or otherwise we shouldn't fill it since this isn't the correct semantics

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/translator.go
Comment (anuraaga) [inline receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/translator.go:22]:
```suggestion
func convertToOCMetrics(prefix string, m ECSMetrics, labelKeys []*metricspb.LabelKey, labelValues []*metricspb.LabelValue, timestamp *timestamppb.Timestamp) []*metricspb.Metric {
```

Comment (anuraaga) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/translator.go:22]: We're actually converting to opencensus metrics here

**Review by anuraaga — COMMENTED** [review]
Left some flyby comments, but basically LGTM with @asuresh4's comments, thanks.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go
Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go]: Removed.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go
Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go]: Yea. I looked into the `dockerstatsreceiver` and it's not setting the value I think. We do have `PullStartedAt` timestamp and I think we can utilize it. However, need more thoughts on this. For now, not setting this value to get the default behavior. Will send a separate PR for this.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go
Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go]: As @anuraaga mentioned, this will be dropped eventually. However, if it does get utilized, we prefer to use `aoc.ecs` for all metrics we are receiving from this receiver.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go
Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go]: Thanks. Added those as Resource Attributes.

File: receiver/awsecscontainermetricsreceiver/go.mod
Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/go.mod]: Updated.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go
Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:32]: Yea, got it. But, for all of our AWS receivers, we are using the receiver name as `service.name`. Because, this field will be utilized by our CW EMFExporter to generate different rules for different receivers. Especially for Container Insights.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/translator.go
Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/translator.go]: Updated.

**Comment by hossain-rayhan** [comment]
Thanks @asuresh4  and @anuraaga thanks for your review. I pushed an update based on the review. Will appreciate another look.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go
Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:33]: Noted. Even if it gets utilized we want to use "aoc.ecs" to differentiate our OT metrics from ECS backend metrics.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go
Comment (bogdandrutu) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go:21]: No, we changed most of the core to use the otlp and internal structs. Completely recommend for new components to avoid oc.

Comment (bogdandrutu) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go:21]: @hossain-rayhan you need to start using `pdata.Metrics`

Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go:21]: Hi @bogdandrutu , before sending the data to next consumer I am using `internaldata.OCToMetrics(md)` to convert our metrics to `pdata.Metrics`. Wondering, isn't that enough like other receivers in the repo or we should strictly get rid of it now?

Comment (bogdandrutu) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go:21]: That is a temporary solution to make progress and not have to change all components once. And decided to use that for some old components that we did not have time to chnage

Comment (anuraaga) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go:21]: @hossain-rayhan Yeah so basically you should only be converting at the last moment when passing down, but here in this sort of receiver-specific logic we want to be using `pdata`, the OTel format. Or we just have to rewrite it right away. We're also having data-model issues because of using the old format (Resource type for example) and we want to make sure the model is right

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go
Comment (anuraaga) [inline receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:31]: Use convention

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/receiver/awsxrayreceiver/internal/translator/aws.go#L21

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go
Comment (anuraaga) [inline receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:30]: Why do we need to prefix the metrics with container? If they have container label, they're container metrics right?

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go
Comment (anuraaga) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:32]: What do you mean by AWS receivers? I think the only one we have is xray, which doesn't do this, and definitely shouldn't since we need to make sure the app's service name is used.

I'm not sure what you mean exactly by the rules, but anyways we can't just fill a semantic convention attribute with something that doesn't follow the spec. If anything, the telemetry.sdk matches closer to what this sort of receiver is doing. @bogdandrutu @tigrannajaryan any suggestion on that?

Also @hossain-rayhan it's important to take a step back and remember what this receiver is here for - it's to translate the container metrics data into the OpenTelemetry format / specification. This is because this data seems useful to users regardless of if they use cloudwatch or not. While we may need some, but hopefully not much, consideration for specific vendors like cloudwatch, that's not the intent here. If you haven't yet, you should go through in detail at least the Resource and Metrics semantics conventions of OTel spec before proceeding and make sure you are aligned with it https://github.com/open-telemetry/opentelemetry-specification/tree/master/specification/resource. That doesn't mean we want to block data that's needed, but it's important to follow the spec as much as possible.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go
Comment (anuraaga) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:33]: I point it out because it will go away - so if you do have an expectation of having it, it won't be there :P But I think if we put the receiver name in something like `telemetry.sdk` than the information should still be preserved, in a way that matches in some sense what our apps send.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go
Comment (asuresh4) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:30]: I'm not sure if this part of the OTel convention but given that other receivers follow this approach, I think we should do the same here for consistency.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go
Comment (mxiamxia) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:32]: This receiver generates ECS container Metrics itself but not receiving any metrics from outside of AOC. For the metrics generated inside the receiver, the idea is to put receiver name in ```service.name``` attribute on these metrics. It's similar to the idea Prometheus receiver uses ```job_name``` as ```service.name``` for metrics it scrapes.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go
Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go:21]: Thanks @bogdandrutu and @anuraaga. I understand we need to use `pdata` eventually. I was planning to move forward with this to meet our internal deadline. We can send a different PR after October 15th I guess. How do you guys feel about it?

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go
Comment (tigrannajaryan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:32]: >I'm not sure what you mean exactly by the rules, but anyways we can't just fill a semantic convention attribute with something that doesn't follow the spec. If anything, the telemetry.sdk matches closer to what this sort of receiver is doing. @bogdandrutu @tigrannajaryan any suggestion on that?

+1. We should not use "service.name" for receiver name. That is not the purpose of "service.name". "service.name" is supposed to describe the source that emits the metrics. Collector is just collecting the metric, it is an intermediary, it is not the source. Nor is "telemetry.sdk" intended for that.

The source that emits the metrics is the container here. If we know the name of the service that runs in the container we should set that. If we don't know we should not record it at all.

I do not know why we want to record the receiver name, perhaps you can clarify the use case. This can then be added as a semantic convention for OpenTelemetry as a whole or just for the Collector and will possibly end up in the "otel" namespace.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go
Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:30]: Yea, I found some other receivers are doing the same like `kubeletstatsreceiver` and `dockerstatsreceiver`.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go
Comment (bogdandrutu) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go:21]: As long as you create an issue and assign to you and @anuraaga I am fine. I trust that you will fix it. I will let @anuraaga make the final call here.

Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go:21]: Issue created: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1122

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go
Comment (anuraaga) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:31]: How about this? `conventions.AttributeServiceName`

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go
Comment (anuraaga) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:32]: @hossain-rayhan If you can give some more detail about this usage that would be great. I think filling in wrong information is blocking this PR, so the easiest way to proceed would be to just remove setting the service name for now and we can figure out a way to handle what you need in a separate PR.

Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/constant.go:32]: Thanks @anuraaga . I am removing it for now as this should not block the receiver. This is more related to the exporter logic as it's being utilized for supporting special customer use cases. If needed I can support it in a separate PR after further discussion.

**Review by anuraaga — APPROVED** [review]
Thanks!

**Comment by hossain-rayhan** [comment]
Hi @bogdandrutu @tigrannajaryan @asuresh4 can we get this merged?

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go
Comment (asuresh4) [inline receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go:35]: nit: `taskResource` would be more accurate. Same with `containerResources` (-> `containerResource`)

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go
Comment (asuresh4) [inline receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go:70]: If the 3rd and 4th parameters to this method are always nil, I would remove those parameters.

**Review by asuresh4 — APPROVED** [review]
LGTM, apart from couple of minor comments.

**Comment by tigrannajaryan** [comment]
I'll merge once the last comments from @asuresh4 are addressed.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go
Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go]: Updated.

File: receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go
Comment (hossain-rayhan) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go:70]: I also thought about it while writing this piece of code. Here, I kept the skeleton ready and the same method can be utilized to set metric labels. In our next PRs, we can just pass the `LabelKeys` and `LabelValues` and we are done. If we really don't utilize, I will remove them.

Comment (asuresh4) [reply receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/accumulator.go:70]: 👍 SGTM

**Pull request merged**
